lib/linkerd2-trace-context/src/layer.rs

```diff
@@ -0,0 +1,186 @@
+{
+    type Service = Stack<M, S>;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        Stack {
+            inner,
+            sink: self.sink.clone(),
+        }
+    }
+}
+
+// === impl Stack ===
+
+impl<T, M, S> tower::Service<T> for Stack<M, S>
+where
+    M: tower::Service<T>,
+    S: Clone,
+{
+    type Response = Service<M::Response, S>;
+    type Error = M::Error;
+    type Future = MakeFuture<M::Future, S>;
+
+    fn poll_ready(&mut self) -> Poll<(), M::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let inner = self.inner.call(target);
+
+        MakeFuture {
+            inner,
+            sink: Some(self.sink.clone()),
+        }
+    }
+}
+
+// === impl MakeFuture ===
+
+impl<F: Future, S> Future for MakeFuture<F, S> {
+    type Item = Service<F::Item, S>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = try_ready!(self.inner.poll());
+        let sink = self.sink.take().expect("poll called after ready");
+        Ok(Async::Ready(Service { inner, sink }))
+    }
+}
+
+// === impl Service ===
+
+impl<Svc, B, S> tower::Service<http::Request<B>> for Service<Svc, S>
+where
+    Svc: tower::Service<http::Request<B>>,
+    S: SpanSink + Clone,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = Either<Svc::Future, SpanFuture<Svc::Future, S>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Svc::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut request: http::Request<B>) -> Self::Future {
+        let mut trace_context = propagation::unpack_trace_context(&request);
+        let mut path = None;
+
+        if let Some(ref mut context) = trace_context {
+            trace!(message = "got trace context", ?context);
+            propagation::increment_span_id(&mut request, context);
+            // If we plan to sample this span, we need to copy the path from
+            // the request before dispatching it to inner.
+            if context.is_sampled() {
+                path = request
+                    .uri()
+                    .path_and_query()
+                    .map(|pq| pq.as_str().to_owned())
+            }
+        }
+
+        let f = self.inner.call(request);
+
+        if let Some(propagation::TraceContext {
+            trace_id,
+            parent_id,
+            flags,
+            span_id: Some(span_id),
+            ..
+        }) = trace_context
+        {
+            if flags.is_sampled() {
+                trace!(message = "span will be sampled", ?span_id);
+                let span = Span {
+                    trace_id,
+                    span_id,
+                    parent_id,
+                    span_name: path.unwrap_or_else(String::new),
+                    start: SystemTime::now(),
+                    // End time will be updated when the span completes.
+                    end: SystemTime::now(),
```

Comment:
Good call.  This actually allowed me to simplify some stuff as well.

Reply:
Nit: If we're going to create spans with an `end` time when they haven't actually ended, we should probably try to avoid calling `SystemTime::now` twice --- this is going to result in a second unnecessary syscall. What about calling `SystemTime::now` once and then cloning that `SystemTime` for the placeholder "end" time? Or, we could use `SystemTime::UNIX_EPOCH` for the placeholder value?